File: src/2024/day8.zig
const std = @import("std");
const utils = @import("utils");
const Point2d = utils.Point2d(i32);

pub fn main() !void {
    // https://adventofcode.com/2024/day/8
    utils.printTitle(2024, 8, "Resonant Collinearity");

    const m = utils.Monitor.init();
    defer m.deinit();
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const ally = gpa.allocator();
    const input = comptime utils.trimInput(@embedFile("input"));

    const p1 = try part1(input, ally);
    utils.printAnswer(1, p1);
    std.debug.assert(276 == p1);

    const p2 = try part2(input, ally);
    utils.printAnswer(2, p2);
    std.debug.assert(991 == p2);
}

const Map = struct {
    frequencies: std.AutoHashMap(u8, std.ArrayList(Point2d)),
    width: i32,
    height: i32,

    pub fn parse(input: []const u8, ally: std.mem.Allocator) !Map {
        var frequencies = std.AutoHashMap(u8, std.ArrayList(Point2d)).init(ally);
        var width: i32 = 0;
        var y: i32 = 0;
        var it = std.mem.tokenizeAny(u8, input, "\r\n");
        while (it.next()) |line| : (y += 1) {
            for (0.., line) |x, c| {
                if (c != '.') {
                    const p = Point2d{ .x = @intCast(x), .y = y };

                    if (frequencies.getPtr(c)) |list| {
                        try list.append(p);
                    } else {
                        var list = std.ArrayList(Point2d).init(ally);
                        try list.append(p);
                        try frequencies.put(c, list);
                    }
                }

                width = @max(width, @as(i32, @intCast(x)) + 1);
            }
        }

        return .{
            .frequencies = frequencies,
            .width = width,
            .height = y,
        };
    }

    pub fn deinit(self: *@This()) void {
        var it = self.frequencies.valueIterator();
        while (it.next()) |l| {
            l.deinit();
        }
        self.frequencies.deinit();
    }

    pub fn contains(self: *const @This(), p: Point2d) bool {
        return p.x >= 0 and p.y >= 0 and p.x < self.width and p.y < self.height;
    }
};

fn part1(input: []const u8, ally: std.mem.Allocator) !i32 {
    var map = try Map.parse(input, ally);
    defer map.deinit();

    var antinodes = std.AutoHashMap(Point2d, void).init(ally);
    defer antinodes.deinit();

    var it = map.frequencies.valueIterator();
    while (it.next()) |frequency| {
        for (0..frequency.items.len) |i| {
            for (i + 1..frequency.items.len) |j| {
                const a = frequency.items[i];
                const b = frequency.items[j];
                const diff = b.subp(a);

                const c = a.subp(diff);
                if (map.contains(c)) {
                    try antinodes.put(c, {});
                }

                const d = b.addp(diff);
                if (map.contains(d)) {
                    try antinodes.put(d, {});
                }
            }
        }
    }

    return @intCast(antinodes.count());
}

fn part2(input: []const u8, ally: std.mem.Allocator) !i32 {
    var map = try Map.parse(input, ally);
    defer map.deinit();

    var antinodes = std.AutoHashMap(Point2d, void).init(ally);
    defer antinodes.deinit();

    var it = map.frequencies.valueIterator();
    while (it.next()) |frequency| {
        for (0..frequency.items.len) |i| {
            for (i + 1..frequency.items.len) |j| {
                var a = frequency.items[i];
                var b = frequency.items[j];
                const diff = b.subp(a);

                while (map.contains(a)) : (a = a.subp(diff)) {
                    try antinodes.put(a, {});
                }

                while (map.contains(b)) : (b = b.addp(diff)) {
                    try antinodes.put(b, {});
                }
            }
        }
    }

    return @intCast(antinodes.count());
}

test "part 1,2" {
    const example =
        \\............
        \\........0...
        \\.....0......
        \\.......0....
        \\....0.......
        \\......A.....
        \\............
        \\............
        \\........A...
        \\.........A..
        \\............
        \\............
    ;
    try std.testing.expectEqual(14, try part1(example, std.testing.allocator));
    try std.testing.expectEqual(34, try part2(example, std.testing.allocator));
}

File: src/2024/day8.rs
#[path = "../utils.rs"]
mod utils;
use itertools::Itertools;
use std::collections::{HashMap, HashSet};
use utils::Point2d;

fn main() {
    // https://adventofcode.com/2024/day/8
    println!("Day 8, 2024: Resonant Collinearity");
    let _m = utils::Monitor::start();

    let p1 = part1(get_input());
    println!("  Part 1: {}", p1);
    assert_eq!(276, p1);

    let p2 = part2(get_input());
    println!("  Part 2: {}", p2);
    assert_eq!(991, p2);
}

struct Map {
    frequencies: HashMap<char, Vec<Point2d>>,
    width: i32,
    height: i32,
}

impl Map {
    pub fn new(input: &str) -> Map {
        let mut width: i32 = 0;
        let mut height: i32 = 0;

        let mut frequencies = HashMap::<char, Vec<Point2d>>::new();
        for (y, line) in input.lines().enumerate() {
            for (x, c) in line.char_indices() {
                if c != '.' {
                    let p = Point2d::new(x as i32, y as i32);
                    if let Some(locations) = frequencies.get_mut(&c) {
                        locations.push(p);
                    } else {
                        frequencies.insert(c, vec![p]);
                    }
                }
                width = x as i32 + 1;
            }

            height = y as i32 + 1;
        }

        Map {
            width: width,
            height: height,
            frequencies: frequencies,
        }
    }

    pub fn contains(&self, p: Point2d) -> bool {
        p.x >= 0 && p.y >= 0 && p.x < self.width && p.y < self.height
    }
}

fn part1(input: &'static str) -> u64 {
    let map = Map::new(input);
    let mut antinodes = HashSet::<Point2d>::new();

    for frequency in map.frequencies.values() {
        for pairs in frequency.iter().combinations(2) {
            let a = **pairs.first().unwrap();
            let b = **pairs.last().unwrap();
            let diff = b - a;

            let c = a - diff;
            if map.contains(c) {
                antinodes.insert(c);
            }

            let d = b + diff;
            if map.contains(d) {
                antinodes.insert(d);
            }
        }
    }

    antinodes.len() as u64
}

fn part2(input: &'static str) -> u64 {
    let map = Map::new(input);
    let mut antinodes = HashSet::<Point2d>::new();

    for frequency in map.frequencies.values() {
        for pairs in frequency.iter().combinations(2) {
            let mut a = **pairs.first().unwrap();
            let mut b = **pairs.last().unwrap();
            let diff = b - a;

            while map.contains(a) {
                antinodes.insert(a);
                a -= diff;
            }

            while map.contains(b) {
                antinodes.insert(b);
                b += diff;
            }
        }
    }

    antinodes.len() as u64
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2024/day8.txt").trim_end()
}

#[cfg(test)]
mod tests {
    fn get_example() -> &'static str {
        indoc::indoc! {"
            ............
            ........0...
            .....0......
            .......0....
            ....0.......
            ......A.....
            ............
            ............
            ........A...
            .........A..
            ............
            ............
        "}
    }

    #[test]
    fn part1() {
        assert_eq!(14, super::part1(get_example()));
    }

    #[test]
    fn part2() {
        assert_eq!(34, super::part2(get_example()));
    }
}
